### PR TITLE
To translate leader pattern use mode that applies on leader element

### PR DESCRIPTION
**Pull request by bertfrees**
@kalaspuffar @PaulRambags I've made it a bit more logical how a leader pattern is translated.

Dotify now uses the mode that applies on `leader` element, that is, the mode that is computed by starting from the `leader` element and taking the `translate` attribute of the first ancestor that has the attribute, or else the specified value for the "mode" option. Before, Dotify took the mode of the first text segment after or before the leader element. This was less logical and made it impossible to specify different modes on the leader pattern and the surrounding text.

### CODE REVIEW

File: src/org/daisy/dotify/formatter/impl/row/LeaderManager.java [hunk 1]
```diff
@@ -9,24 +9,29 @@
...
 import java.util.logging.Logger;
+import java.util.Optional;
 
 /**
  * TODO: Write java doc.
  */
 class LeaderManager {
     private static final Logger logger = Logger.getLogger(LeaderManager.class.getCanonicalName());
     private Deque<Leader> leaders;
+	private Deque<Optional<String>> leaderModes;
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/row/LeaderManager.java]: fixed

File: src/org/daisy/dotify/api/formatter/BlockContentBuilder.java [hunk 2]
```diff
@@ -27,9 +27,10 @@
      * Insert a leader at the current position in the flow.
      *
      * @param leader the leader to insert
+     * @param props the TextProperties including the mode to translate the pattern
      * @throws IllegalStateException if the current state does not allow this call to be made
      */
-    public void insertLeader(Leader leader);
+    public void insertLeader(Leader leader, TextProperties props);
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply src/org/daisy/dotify/api/formatter/BlockContentBuilder.java:33]: I chose to use the same name as in the `addChars` method below.

File: src/org/daisy/dotify/formatter/impl/core/FormatterCoreImpl.java [hunk 3]
```diff
@@ -335,11 +335,11 @@ public void insertAnchor(String ref) {
     }
 
     @Override
-    public void insertLeader(Leader leader) {
+    public void insertLeader(Leader leader, TextProperties p) {
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/core/FormatterCoreImpl.java:338]: I chose to use the same name as in the `addChars` method below.

File: src/org/daisy/dotify/formatter/impl/core/TableOfContentsImpl.java [hunk 4]
```diff
@@ -220,9 +220,9 @@ public void insertAnchor(String ref) {
     }
 
     @Override
-    public void insertLeader(Leader leader) {
+    public void insertLeader(Leader leader, TextProperties props) {
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/core/TableOfContentsImpl.java:223]: I chose to use the same name as in the `addChars` method below.

**Review by PaulRambags — APPROVED** [review]

File: src/org/daisy/dotify/formatter/impl/row/LeaderManager.java [hunk 5]
```diff
@@ -9,24 +9,29 @@
...
 import java.util.logging.Logger;
+import java.util.Optional;
 
 /**
  * TODO: Write java doc.
  */
 class LeaderManager {
     private static final Logger logger = Logger.getLogger(LeaderManager.class.getCanonicalName());
     private Deque<Leader> leaders;
+    private Deque<Optional<String>> leaderModes;
```
Comment (kalaspuffar) [inline src/org/daisy/dotify/formatter/impl/row/LeaderManager.java:20]: Could these queues get out of sync, and if so isn't it more appropriate to put the mode as a member of the leader?

**Review by kalaspuffar — APPROVED** [review]

File: src/org/daisy/dotify/formatter/impl/row/LeaderManager.java (hunk 5 above)
Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/row/LeaderManager.java:20]: Normally not, but the mode could be included in the `Leader` object. Or in fact, the `pattern` field could be set to the translated pattern. Yes, I think that's a nicer solution.

Comment (bertfrees) [reply src/org/daisy/dotify/formatter/impl/row/LeaderManager.java]: Addressed this in my latest commit.

**Comment by bertfrees** [comment]
Please check my latest commit. (It's OK to squash it when you merge.)

**Comment by kalaspuffar** [comment]
PR approved, running regression tests.

**Comment by kalaspuffar** [comment]
Regression tests passed, merging.

**Pull request merged**